cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs

```diff
@@ -665,26 +600,34 @@ txFeatureMismatch era feature =
     left (ShelleyTxCmdTxFeatureMismatch (anyCardanoEra era) feature)
 
 validateTxIns
-  :: forall era.
-     CardanoEra era
-  -> [(TxIn, Maybe (ScriptWitnessFiles WitCtxTxIn))]
-  -> ExceptT ShelleyTxCmdError IO
-             [(TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn era))]
-validateTxIns era = mapM convert
+  :: [(TxIn, Maybe (ScriptWitness WitCtxTxIn era))]
+  -> [(TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn era))]
+validateTxIns = map convert
  where
    convert
-     :: (TxIn, Maybe (ScriptWitnessFiles WitCtxTxIn))
-     -> ExceptT ShelleyTxCmdError IO
-                (TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn era))
-   convert (txin, mScriptWitnessFiles) =
-     case mScriptWitnessFiles of
-       Just scriptWitnessFiles -> do
-         sWit <- createScriptWitness era scriptWitnessFiles
-         return ( txin
-                , BuildTxWith $ ScriptWitness ScriptWitnessForSpending sWit
-                )
-       Nothing -> return (txin, BuildTxWith $ KeyWitness KeyWitnessForSpending)
-
+     :: (TxIn, Maybe (ScriptWitness WitCtxTxIn era))
+     -> (TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn era))
+   convert (txin, mScriptWitness) =
+     case mScriptWitness of
+       Just sWit ->
+         (txin , BuildTxWith $ ScriptWitness ScriptWitnessForSpending sWit)
+       Nothing ->
+         (txin, BuildTxWith $ KeyWitness KeyWitnessForSpending)
+
+readScriptWitnessFiles
+  :: CardanoEra era
+  -> [(TxIn, Maybe (ScriptWitnessFiles WitCtxTxIn))]
+  -> ExceptT ShelleyTxCmdError IO [(TxIn, Maybe (ScriptWitness WitCtxTxIn era))]
+readScriptWitnessFiles _ [] = return []
+readScriptWitnessFiles era ((txin, mScriptWitnessFiles) : rest) = do
+  case  mScriptWitnessFiles of
+    Nothing -> do
+      remaining <- readScriptWitnessFiles era rest
+      return $ (txin, Nothing) : remaining
+    Just sWitFile -> do
+      sWit <- createScriptWitness era sWitFile
+      remaining <- readScriptWitnessFiles era rest
+      return $ (txin, Just sWit) : remaining
```

Comment:
Is there a reason to prefer explicit recursion over something like

```suggestion
readScriptWitnessFiles era = mapM go
  where
    go (txin, mScriptWitnessFiles) = do
      mSWit <- mapM (readScriptWitnessFiles era) mScriptWitnessFiles
      return (txin, mSWit)
```

Reply:
I agree! :+1:

Reply:
Refactored :+1: